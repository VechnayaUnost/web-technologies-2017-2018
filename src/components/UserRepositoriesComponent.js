import React from 'react';
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { Link } from 'react-router-dom';
import { fetchUserReposRequest } from '../actions'

class UserRepositoriesComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            currentRepoType: 'All',
            currentRepoLang: 'All'
        }
    }

    componentDidMount() {
        this.props.fetchUserReposRequest(this.props.userLogin);
    }

    getUniqueLanguages(arr) {
        let tempObj = {};

        for (let i = 0; i < arr.length; i++) {
            let key = arr[i].language
            tempObj[key] = true;
        }

        return Object.keys(tempObj);
    }

    changeRepoType(e) {
        this.setState({
            currentRepoType: e.target.value
        })
        console.log(e.target.value)
    }

    changeRepoLang(e) {
        this.setState({
            currentRepoLang: e.target.value
        })
        console.log(e.target.value)
    }

    filterRepos(repos) {
        console.log(this.state);
        console.log(repos);
        const firstResult = repos.filter(repo => {
            if (this.state.currentRepoLang === 'All') {
                console.log(this.state.currentRepoLang);
                console.log(repo);
                return repo
            } else {
                console.log(this.state.currentRepoLang);
                console.log(repo);
                return repo.language === this.state.currentRepoLang
            }
        })
        const secondResult = firstResult.filter(repo => {
            if (this.state.currentRepoType === 'All') {
                console.log(this.state.currentRepoType);
                console.log(repo);
                return repo
            } else {
                switch(this.state.currentRepoType) {
                    case 'Forks': {
                        console.log(this.state.currentRepoType);
                        console.log(repo);
                        return repo.fork
                    }
                    case 'Archived': {
                        console.log(this.state.currentRepoType);
                        console.log(repo);
                        return repo.archived
                    }
                    default: return;
                }
            }
        })
        console.log(secondResult);
        return secondResult;
    }

    render() {
        console.log(this.props.repos);
        const repos = [];
        for (const repo in this.props.repos) {
            if (this.props.repos.hasOwnProperty(repo)) {
                const element = this.props.repos[repo];
                repos.push(element);
            }
        }
        const reposLangs = this.getUniqueLanguages(repos);
        console.log(reposLangs);
        const filteredRepos = this.filterRepos(repos);

        return (
            this.props.userLogin === '' || this.props.userLogin === undefined
                ?
                <div>
                    <ul>
                        <li>
                            <Link to="/user/followers">User followers</Link>
                        </li>
                        <li>
                            <Link to="/user/repos">User repositories</Link>
                        </li>
                    </ul>

                    <h2>Repositories</h2>
                    <h3>Empty</h3>
                </div>
                :
                <div>
                    <ul>
                        <li>
                            <Link to="/user/followers">User followers</Link>
                        </li>
                        <li>
                            <Link to="/user/repos">User repositories</Link>
                        </li>
                    </ul>
                    <h2>Repositories</h2>
                    <select className="repos-type" value={this.state.currentRepoType} onChange={this.changeRepoType.bind(this)}>
                        <option value="All">All</option>
                        <option value="Forks">Forks</option>
                        <option value="Archived">Archived</option>
                    </select>
                    <select className="repos-lang" value={this.state.currentRepoLang} onChange={this.changeRepoLang.bind(this)}>
                        <option value="All">All</option>
                        {reposLangs.map(repoLang => <option value={repoLang}>{repoLang}</option>)}
                    </select>
                    {filteredRepos.map(repo =>
                        <div>
                            <h3>{repo.name}</h3>
                            <span>{repo.description}</span>
                            <br />
                            <span>Stars: {repo.stargazers_count}</span>
                            <br />
                            <span>Watchers: {repo.watchers_count}</span>
                        </div>
                    )}
                </div>
        )
    }
}

const mapStateToProps = (state) => {
    console.log(state)
    console.log(state.userRepos)
    return {
        userLogin: state.userInfo.login,
        repos: state.userRepos
    };
};

const mapDispatchToProps = dispatch => bindActionCreators({
    fetchUserReposRequest,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(UserRepositoriesComponent);