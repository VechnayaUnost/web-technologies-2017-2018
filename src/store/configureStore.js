import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware, { END } from 'redux-saga';

import rootReducer from '../reducers';
import rootSaga from '../sagas';

export default (initialState) => {
    const sagaMiddleware = createSagaMiddleware();
    const store = createStore(rootReducer, initialState, applyMiddleware(sagaMiddleware));

    sagaMiddleware.run(rootSaga);
    store.runSaga = () => sagaMiddleware.run(rootSaga);
    store.close = () => store.dispatch(END);

    return store;
};