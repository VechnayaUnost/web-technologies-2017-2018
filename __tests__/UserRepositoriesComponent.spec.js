import React from 'react';
import UserRepositoriesComponent from '../src/components/UserRepositoriesComponent';
import { shallow, mount, configure } from 'enzyme';
import { shallowToJson, mountToJson } from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

configure({ adapter: new Adapter() });

describe('UserRepositories component test', () => {
    const mockStore = configureStore([ thunk ]);
    const initialState = {
        userInfo: {
            login: 'GoGoPandora'
        },
        userRepos: [
            {
                name: 'AvatarSkeleton',
                description: 'Creates Avatar skeleton with some preinstalled options',
                stargazers_count: 321,
                watchers_count: 123,
                language: 'C#',
                fork: true,
                archived: false
            },
            {
                name: 'AvatarMountConnection',
                description: 'Avatar mount connection is easier with this tool',
                stargazers_count: 543,
                watchers_count: 345,
                language: 'Java',
                fork: false,
                archived: true
            },
            {
                name: 'SoldiersBarracks',
                description: 'Barrack project for human soldiers',
                stargazers_count: 2,
                watchers_count: 1,
                language: 'Paper&Pen',
                fork: false,
                archived: false
            }
        ]
    }

    let store;

    beforeEach(() => {
        store = mockStore(initialState);
    })

    it('UserRepositories component rendered succesfully', () => {
        const userRepositories = mount(
            <BrowserRouter>
                <UserRepositoriesComponent store={store} />
            </BrowserRouter>
        );
        expect(mountToJson(userRepositories)).toMatchSnapshot();
    });

    it('Switch filter', () => {
        const userRepositories = mount(
            <BrowserRouter>
                <UserRepositoriesComponent store={store} />
            </BrowserRouter>
        );
        expect(userRepositories.find('.repos-type').props().value).toBe('All');
        expect(userRepositories.find('.repos-lang').props().value).toBe('All');
        userRepositories.find('.repos-type').simulate('change', { target: { value: 'Forks' } });
        userRepositories.find('.repos-lang').simulate('change', { target: { value: 'C#' } });
        expect(userRepositories.find('.repos-type').props().value).toBe('Forks');
        expect(userRepositories.find('.repos-lang').props().value).toBe('C#');
        userRepositories.find('.repos-type').simulate('change', { target: { value: 'Archived' } });
        expect(userRepositories.find('.repos-type').props().value).toBe('Archived');
        userRepositories.find('.repos-type').simulate('change', { target: { value: 'SomethingElse' } });
        expect(userRepositories.find('.repos-type').props().value).toBe('SomethingElse');
    });
    
    it('UserRepositories rendered without user login', () => {
        const fakeInitialState = {
            userInfo: {
                login: ''
            },
            userRepos: [
                {
                    name: 'AvatarSkeleton',
                    description: 'Creates Avatar skeleton with some preinstalled options',
                    stargazers_count: 321,
                    watchers_count: 123
                }
            ]
        }
        const fakeStore = mockStore(fakeInitialState);
        const userRepositories = mount(
            <BrowserRouter>
                <UserRepositoriesComponent store={fakeStore} />
            </BrowserRouter>
        );
        expect(userRepositories.prop('userLogin')).toBe(undefined);
        expect(mountToJson(userRepositories)).toMatchSnapshot();
    })
});