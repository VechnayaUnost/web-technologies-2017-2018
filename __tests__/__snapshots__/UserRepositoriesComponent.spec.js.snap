// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepositories component test UserRepositories component rendered succesfully 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(UserRepositoriesComponent)
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <UserRepositoriesComponent
        fetchUserReposRequest={[Function]}
        repos={
          Array [
            Object {
              "archived": false,
              "description": "Creates Avatar skeleton with some preinstalled options",
              "fork": true,
              "language": "C#",
              "name": "AvatarSkeleton",
              "stargazers_count": 321,
              "watchers_count": 123,
            },
            Object {
              "archived": true,
              "description": "Avatar mount connection is easier with this tool",
              "fork": false,
              "language": "Java",
              "name": "AvatarMountConnection",
              "stargazers_count": 543,
              "watchers_count": 345,
            },
            Object {
              "archived": false,
              "description": "Barrack project for human soldiers",
              "fork": false,
              "language": "Paper&Pen",
              "name": "SoldiersBarracks",
              "stargazers_count": 2,
              "watchers_count": 1,
            },
          ]
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
        userLogin="GoGoPandora"
      >
        <div>
          <ul>
            <li>
              <Link
                replace={false}
                to="/user/followers"
              >
                <a
                  href="/user/followers"
                  onClick={[Function]}
                >
                  User followers
                </a>
              </Link>
            </li>
            <li>
              <Link
                replace={false}
                to="/user/repos"
              >
                <a
                  href="/user/repos"
                  onClick={[Function]}
                >
                  User repositories
                </a>
              </Link>
            </li>
          </ul>
          <h2>
            Repositories
          </h2>
          <select
            className="repos-type"
            onChange={[Function]}
            value="All"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="Forks"
            >
              Forks
            </option>
            <option
              value="Archived"
            >
              Archived
            </option>
          </select>
          <select
            className="repos-lang"
            onChange={[Function]}
            value="All"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="C#"
            >
              C#
            </option>
            <option
              value="Java"
            >
              Java
            </option>
            <option
              value="Paper&Pen"
            >
              Paper&Pen
            </option>
          </select>
          <div>
            <h3>
              AvatarSkeleton
            </h3>
            <span>
              Creates Avatar skeleton with some preinstalled options
            </span>
            <br />
            <span>
              Stars: 
              321
            </span>
            <br />
            <span>
              Watchers: 
              123
            </span>
          </div>
          <div>
            <h3>
              AvatarMountConnection
            </h3>
            <span>
              Avatar mount connection is easier with this tool
            </span>
            <br />
            <span>
              Stars: 
              543
            </span>
            <br />
            <span>
              Watchers: 
              345
            </span>
          </div>
          <div>
            <h3>
              SoldiersBarracks
            </h3>
            <span>
              Barrack project for human soldiers
            </span>
            <br />
            <span>
              Stars: 
              2
            </span>
            <br />
            <span>
              Watchers: 
              1
            </span>
          </div>
        </div>
      </UserRepositoriesComponent>
    </Connect(UserRepositoriesComponent)>
  </Router>
</BrowserRouter>
`;

exports[`UserRepositories component test UserRepositories rendered without user login 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "blank",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(UserRepositoriesComponent)
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <UserRepositoriesComponent
        fetchUserReposRequest={[Function]}
        repos={
          Array [
            Object {
              "description": "Creates Avatar skeleton with some preinstalled options",
              "name": "AvatarSkeleton",
              "stargazers_count": 321,
              "watchers_count": 123,
            },
          ]
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
        userLogin=""
      >
        <div>
          <ul>
            <li>
              <Link
                replace={false}
                to="/user/followers"
              >
                <a
                  href="/user/followers"
                  onClick={[Function]}
                >
                  User followers
                </a>
              </Link>
            </li>
            <li>
              <Link
                replace={false}
                to="/user/repos"
              >
                <a
                  href="/user/repos"
                  onClick={[Function]}
                >
                  User repositories
                </a>
              </Link>
            </li>
          </ul>
          <h2>
            Repositories
          </h2>
          <h3>
            Empty
          </h3>
        </div>
      </UserRepositoriesComponent>
    </Connect(UserRepositoriesComponent)>
  </Router>
</BrowserRouter>
`;
